estimation.py

```diff
@@ -0,0 +1,212 @@
+
+        if not active_fake_mode():
+            whole_model.init_weights()
+
+        # build optimizer after applying parallelisms to the model
+        optimizers = build_optimizers(model_parts, job_config)
+        lr_schedulers = get_lr_schedulers(optimizers.optimizers, job_config)
+
+        for model in model_parts:
+            model.train()
+        logger.info(f"Vocab size: {model_config.vocab_size}")
+        # Create a dummy batch instead of loading from a dataset
+        batch = (
+            torch.randint(
+                0,
+                model_config.vocab_size,
+                (job_config.training.batch_size, model_config.max_seq_len),
+                device="cuda",
+            ),
+            torch.randint(
+                0,
+                model_config.vocab_size,
+                (job_config.training.batch_size, model_config.max_seq_len),
+                device="cuda",
+            ),
+        )
+        fsdp_memtracker = FSDPMemTracker(mod=whole_model, optm=optimizers.optimizers[0])
+        fsdp_memtracker.track_inputs(batch)
+
+        with fsdp_memtracker:
+            for iter_idx in range(2):
+                input_ids, labels = batch
+                # train step
+                with loss_parallel_ctx():
+                    pred = whole_model(input_ids)
+                    loss = loss_fn(pred, labels)
+                    del pred
+                    loss.backward()
+
+                # clip gradients
+                for model in model_parts:
+                    torch.nn.utils.clip_grad_norm_(
+                        model.parameters(), job_config.training.max_norm, foreach=True
+                    )
+                # optimizer step
+                optimizers.step()
+                lr_schedulers.step()
+                optimizers.zero_grad()
+                print(f"Peak Memory at iter: {iter_idx}")
+                fsdp_memtracker.display_snapshot("peak", units="MiB", tabulate=True)
+                if iter_idx == 0:
+                    fsdp_memtracker.reset_mod_stats()  # iter 0 does not have optimizer state
+                gc.collect(1)
+
+        fsdp_memtracker.display_modulewise_snapshots(
+            depth=3, units="MiB", tabulate=True
+        )
+        mem_stats = torch.cuda.memory_stats()
+        peak_active = mem_stats["active_bytes.all.peak"]
+        peak_reserved = mem_stats["reserved_bytes.all.peak"]
+        num_retries = mem_stats["num_alloc_retries"]
+        dev = torch.device(torch.cuda.current_device())
+        tracker_peak = fsdp_memtracker.get_tracker_snapshot("peak")[dev]["Total"]
+        gib = 1024**3
+        print(
+            f"peak active: {peak_active / gib} GiB | peak reserved:"
+            f" {peak_reserved / gib} GiB | num_retries: {num_retries}"
+        )
+        print(f"Tracker Max: {tracker_peak / gib} GiB")
+        if job_config.estimate.mode == "real":
+            print(f"Tracker Accuracy: {tracker_peak/peak_active}")
```

Comment:
Would `peak_active` absolutely never be 0? It sounds safer if we explicit check.

Reply:
It will never be zero in "real" mode and it will always be zero in "fake" mode

Reply:
OK, how about let's do an assertion before this line?